shared/src/main/scala/Parser.scala

```diff
@@ -1,494 +1,430 @@
+    * containing whatever it parsed
+    */
+  type ParserRet[T] = Either[VyxalCompilationError, T]
 
-  def number: Parser[AST] =
-    accept(
-      "number",
-      { case VyxalToken.Number(value) => AST.Number(VNum.from(value)) }
-    )
+  private def toValidName(name: String): String =
+    name.filter(_.isLetterOrDigit).dropWhile(!_.isLetter)
 
-  def compositeNilad: Parser[AST] =
-    accept(
-      "number",
-      { case VyxalToken.CompositeNilad(value) =>
-        if (value.length == 1) {
-          val temp = VyxalParser.parseWithoutProcessing(value)
-          temp match {
-            case Right(ast) => ast
-            case Left(error) =>
-              throw RuntimeException(s"Error while parsing $value: $error")
+  /** The parser takes a list of tokens and performs two sweeps of parsing:
+    * structures + arity grouping and then modifiers. The first sweep deals
+    * directly with the token list provided to the parser, and leaves its
+    * results in a stack of ASTs (the stack data type is used because it means
+    * that arity grouping is simply popping previous ASTs until a niladic state
+    * is reached). The second sweep takes the stack of ASTs and applies the
+    * logic of modifier grouping, placing its result in a single Group AST.
+    *
+    * @param topLevel
+    *   Whether this is the topmost call to parse, in which case it should
+    *   remove any `StructureAllClose`s left behind by `parseStructure`
+    */
+  private def parse(
+      program: Queue[VyxalToken],
+      topLevel: Boolean = false
+  ): ParserRet[AST] = {
+    val asts = Stack[AST]()
+    // Convert the list of tokens to a queue so that ASTs like Structures can
+    // freely take as many tokens as they need without needing to worry about
+    // changing indexes like a for-loop would require.
+
+    // Begin the first sweep of parsing
+
+    while (program.nonEmpty && !isCloser(program.front)) {
+      (program.dequeue(): @unchecked) match {
+        // Numbers, strings and newlines are trivial, and are simply evaluated
+        case VyxalToken.Number(value) =>
+          asts.push(AST.Number(VNum.from(value)))
+        case VyxalToken.Str(value) => asts.push(AST.Str(value))
+        case VyxalToken.Newline    => asts.push(AST.Newline)
+        case VyxalToken.StructureOpen(open) =>
+          parseStructure(open, program) match {
+            case Right(ast) => asts.push(ast)
+            case l          => return l
           }
-        } else {
-          val temp = value.map(tok => {
-            val inner = VyxalParser.parse(List(tok))
-            inner match {
-              case Right(ast) => ast
-              case Left(error) =>
-                throw RuntimeException(s"Error while parsing $tok: $error")
+          if (
+            topLevel && program.nonEmpty && program.front == VyxalToken.StructureAllClose
+          ) {
+            program.dequeue()
```

Comment:
This can just be setting the structure depth to 0 and placing a close on the top of the queue

Reply:
In a previous version of the structure parsing logic, I was using an integer variable to keep track of how many structures the program was in.

It seems to be gone now though

Reply:
It's all good I'll just have to make sure I give everything a good look over so I know I understand it when I have the chance